contrib/implements/README.md

```diff
@@ -0,0 +1,49 @@
+
+# The 'implements' tool
+
+implements is a small-ish tool created to compare the Ceph C APIs with
+go-ceph implmeents.
+
+## Build
+
+In the go-ceph repository run `make implmeents` to create a standalone
+binary for the `implements` cli tool.
+
+## Run
+
+```
+./implements [--verbose] [--json] [--list] [--annotate] <DIR> [pkg...]
+```
+
+The --verbose option causes verbose details about the source scan to be
+printed.
+
+The tool can produce either plain-text output, or JSON with the --json option.
+
+The --list option produces a list of all covered and missing functions from
+the Ceph library.
+
+The --annotate option adds information about each function to the listing.
```

Comment:
you're adding `documented`, maybe also add `undocumented` so that it is easier to find the problems? Or some other way... Not mark them as `Covered`, but a 3rd state?

Reply:
I think there are many cases were a function will be used without explicit documentation. It may be useful to know which ones are that way, but that can just be seen that they're called but not documented. I'm not really keen on creating a new category for that.

Reply:
There might be cases where `called` (without `documented`) makes sense, but I hope the majority would be `called documented`. Reducing the number of calls that are undocumented should be a goal, similar to reducing `missing`.

Have you thought about moving called+undocumented functions to `missing (undocumented)` instead of having them added to `covered`?

Reply:
> There might be cases where `called` (without `documented`) makes sense, but I hope the majority would be `called documented`. 

Wholeheartedly agree.

Reducing the number of calls that are undocumented should be a goal, similar to reducing `missing`.

Agree.

> 
> Have you thought about moving called+undocumented functions to `missing (undocumented)` instead of having them added to `covered`?

I suspect this will be the sticking point. Would you prefer if the top-level term used was "found" or "implemented" rather than "covered"?

As for quickly finding stuff that is called but not documented (in text output) a quick:
`./implements --list --annotate | grep called | grep -v documented`

I could drop `--annotate` and just make it always annotate with --list. Originally, I wanted the options to match the python apicompare.py. I was diffing the output for both to make sure the go version was working as I meant it to. However, I'm pretty sure that the new tool is feature complete as far as that's concerned now.

Reply:
My preference would probably go to "found" instead of "covered". Adding the annotations with `--list` is a better default, as that is something that will be use most, I assume.

If we want the percentage of coverage, we would want to know which functions are implemented completely, and that includes documentation for those features. So, in my opinion, the percentage is currently wrong, as it counts called functions to the "implemented" part, which is not complete for everything yet.

Reply:
Phrased differently, I expect a list of "done" (implemented) and "incomplete/missing". Using `grep` is certainly an option, but it would be good to have this as the goal/output of the tool.

Reply:
> My preference would probably go to "found" instead of "covered". Adding the annotations with `--list` is a better default, as that is something that will be use most, I assume.
> 

I'll make the change to print "found" and drop annotate and just do annotations by default.

> If we want the percentage of coverage, we would want to know which functions are implemented completely, and that includes documentation for those features.

I'm happy to add more stats in the future.

> So, in my opinion, the percentage is currently wrong, as it counts called functions to the "implemented" part, which is not complete for everything yet.
> Phrased differently, I expect a list of "done" (implemented) and "incomplete/missing". Using `grep` is certainly an option, but it would be good to have this as the goal/output of the tool.

To my "incrementalist" sort of thinking I see having something in the code base, even if not explicitly documented, as better than nothing. Even significantly better than nothing. I also don't like the idea of "punishing" ourselves for older code that was not following the policy of adding "Implements" to function calls. There's also the matter of higher-level / compound functions that make use of multiple ceph api calls. I would not like to document them all under implements because that then asks the user to understand how the calls interact *in* our code. The thinly layered one-to-one type calls are the ones that should carry the "Implements" lines.

To me that all adds up to counting undocumented functions as found (or whatever we call it) and not creating a secondary category. But use the annotations to fill in the gaps in documentation over time. We may never see eye-to-eye on this exact point. I'd prefer to go with the current approach with the tweaks mentioned above, for now at least. Is that acceptable?

Reply:
Yes, I agree, so it is fine with me.